### PR TITLE
Allow TimeSeries to have offset attribute

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
## Motivation

Analogous to but ultimately independent of the `MeasurementData` (subclass of `VectorData`) changes upstream (https://github.com/hdmf-dev/hdmf/pull/688), it would be great if NWB supported an `offet` in addition to a `conversion` factor for `TimeSeries` objects.

To convert the `TimeSeries` `data` field to the specified `units`, one would perform the operation

```python
data_in_units = data_in_storage * conversion + offset
```

To-Do:
- [x] Make a sister PR with the `nwb-schema`: https://github.com/NeurodataWithoutBorders/nwb-schema/pull/494

## How to test the behavior?
Added tests that the attribute is propagated.

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR clearly describes the problem and the solution?
- [X] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?

### CODE REVIEW

**Comment by oruebel** [comment]
Not that this is something we should do in this PR, but this is another argument for wrapping datasets to allow us to apply these transformations directly rather than putting the burden on the user to transform the data.

**Comment by CodyCBakerPhD** [comment]
> Not that this is something we should do in this PR, but this is another argument for wrapping datasets to allow us to apply these transformations directly rather than putting the burden on the user to transform the data.

@oruebel That is a good point; our approach so far (for `ecephys` anyway) has been to rely on external APIs such as SpikeInterface (https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/baserecording.py#L110-L114) to handle this transformation but the lack of proper offsets within NWBFiles has required awkward workarounds to get it to work properly.

A core functionality for allowing it to work similarly at the `NWBHDF5IO` level would be useful for people not aware and familiar of those external tools, however; we'll have to explore and discuss some possibilities for that design (I remember hearing something about `h5py` filters that can work in that way?).

File: src/pynwb/base.py [hunk 1]
```diff
@@ -111,7 +112,12 @@ class TimeSeries(NWBDataInterface):
              'doc': 'The smallest meaningful difference (in specified unit) between values in data', 'default': -1.0},
             {'name': 'conversion', 'type': (str, 'float'),
              'doc': 'Scalar to multiply each element in data to convert it to the specified unit', 'default': 1.0},
-
+            {'name': 'offset', 'type': (str, 'float'),
```
Comment (rly) [inline src/pynwb/base.py]: I do not think string should be allowed here. Or for resolution and conversion. That is surely a bug.

File: tests/unit/test_base.py [hunk 2]
```diff
@@ -92,9 +92,24 @@ def test_init_data(self):
...
+    def test_init_conversion(self):
+        dat = [0, 1, 2, 3, 4]
+        conversion = 2.1
+        ts = TimeSeries('test_ts', dat, 'volts', timestamps=[0.1, 0.2, 0.3, 0.4], conversion=conversion)
+        self.assertIs(ts.data, dat)
+        self.assertEqual(ts.conversion, conversion)
+
+    def test_init_offset(self):
+        dat = [0, 1, 2, 3, 4]
+        offset = 0.0
```
Comment (rly) [inline tests/unit/test_base.py]: I would test with a non-default value.

Comment (CodyCBakerPhD) [inline tests/unit/test_base.py]:
```suggestion
        offset = 1.2
```

Comment (CodyCBakerPhD) [reply tests/unit/test_base.py]: Fantastic idea

File: src/pynwb/base.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/pynwb/base.py]: Fixed! I did think that was a bit strange, lol.

**Comment by CodyCBakerPhD** [comment]
@rly Any idea why the tests are failing? They pass for me locally

**Comment by rly** [comment]
The nwb-schema submodule needed to be updated to point to the latest commit on nwb-schema's dev branch. 

miniconda37 fails due to a known package dependency issue.
gallery tests fail because as of 4 days ago, there is an issue with importing allensdk which uses the jinja2 which uses now deprecated functionality from markupsafe.

**Comment by CodyCBakerPhD** [comment]
@rly Hey, sorry, I know your busy - any ETA on when this will be available? It's already in the schema, and is a highly requested feature for all the formats (EDF/intan) that require it for optimal performance.

**Comment by rly** [comment]
Thanks for the ping on this PR. This looks good to go but ideally we should make a new release of the schema first. See https://github.com/NeurodataWithoutBorders/nwb-schema/pull/511 for progress on that.

**Comment by rly** [comment]
Actually, it's almost good to go. "offset" needs to be added to the constructor docval of every subclass of TimeSeries, like for "conversion", e.g., https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/src/pynwb/ecephys.py#L78

**Comment by CodyCBakerPhD** [comment]
> Actually, it's almost good to go. "offset" needs to be added to the constructor docval of every subclass of TimeSeries, like for "conversion", e.g., https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/src/pynwb/ecephys.py#L78

Oh man, those docvals don't get inherited?

**Comment by CodyCBakerPhD** [comment]
Ohh, I see, I just have to go in and add it to each

```python
            *get_docval(TimeSeries.__init__, 'conversion', 'resolution', 'timestamps', 'starting_time', 'rate',
                        'comments', 'description', 'control', 'control_description'))
```

type of line, like so

```python
            *get_docval(TimeSeries.__init__, 'conversion', 'resolution', 'timestamps', 'starting_time', 'rate',
                        'comments', 'description', 'control', 'control_description', 'offset'))
```

right?

Should I add a new test for this for each subclass? @rly

**Comment by CodyCBakerPhD** [comment]
@rly OK I think I updated all of the docvals, using all those ones we had to change in the schema https://github.com/NeurodataWithoutBorders/nwb-schema/pull/494/files# as a reference of what needed propagation.

Let me know if you think I missed any.

**Comment by rly** [comment]
> Should I add a new test for this for each subclass? @rly

No need

**Comment by rly** [comment]
I think you missed a few such as `RoiResponseSeries`. If you go to the "TimeSeries" bullet point in [the nwb-schema docs](https://nwb-schema.readthedocs.io/en/latest/format.html#type-hierarchy) you'll see a type hierarchy of all types with `TimeSeries` as an ancestor.

Please also revert the changes to ImagingPlane (see https://github.com/NeurodataWithoutBorders/nwb-schema/pull/513)

**Comment by CodyCBakerPhD** [comment]
> Please also revert the changes to ImagingPlane (see https://github.com/NeurodataWithoutBorders/nwb-schema/pull/513)

Done

> I think you missed a few such as RoiResponseSeries. If you go to the "TimeSeries" bullet point in [the nwb-schema docs](https://nwb-schema.readthedocs.io/en/latest/format.html#type-hierarchy) you'll see a type hierarchy of all types with TimeSeries as an ancestor.

Oops, sorry about that. Guess basing it off the schema changes didn't account for all the ones that don't need it overridden. I've gone through the list on the readthedocs schema view and added it to all except: `IntervalSeries` and `AnnotationSeries` since they also don't take `conversion`, the necessary compliment to this.

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
Looks good to me. Thanks!

**Pull request merged**